### PR TITLE
test(replica): volume metafile deleted or empty

### PR DESCRIPTION
**Pull request by mantissahz**
Try to rebuild the volume metafile when attaching a volume. Add test case to delete or empty the vollume meta file and then It should still work fine when the volume attached.

longhorn/longhorn#4846
https://github.com/longhorn/longhorn/issues/4964
[PR775](https://github.com/longhorn/longhorn-engine/pull/775)

### CODE REVIEW

**Review by innobead — CHANGES_REQUESTED** [review]
Can we add another case to delete volume meta when the volume is attached and having some IO?

**Review by innobead — COMMENTED** [review]
For each step, please add a comment at the beginning of the step, and this will make the test case more readable.

File: manager/integration/tests/test_basic.py [hunk 1]
```diff
@@ -3642,6 +3642,165 @@ def test_volume_toomanysnapshots_condition(client, core_api, volume_name):  # NO
...
+    test_pod_name = "pod-" + volume_name
+    volume_size = str(1 * Gi)
+
+    csi_pv['metadata']['name'] = test_pv_name
+    csi_pv['spec']['csi']['volumeHandle'] = volume_name
+    csi_pv['spec']['capacity']['storage'] = volume_size
+    pvc['metadata']['name'] = test_pvc_name
+    pvc['spec']['volumeName'] = test_pv_name
+    pvc['spec']['resources']['requests']['storage'] = volume_size
+    pvc['spec']['storageClassName'] = ''
```
Comment (innobead) [inline manager/integration/tests/test_basic.py:3683]: Also, try to see if any useful function in common.py instead of duplicating same operations. ex: `prepare_pod_with_data_in_mb`

You can refer to other cases.

Comment (mantissahz) [reply manager/integration/tests/test_basic.py]: Replaced by `prepare_pod_with_data_in_mb`, and some snapshot checks. Thanks

File: manager/integration/tests/test_basic.py [hunk 2]
```diff
@@ -3642,6 +3643,127 @@ def test_volume_toomanysnapshots_condition(client, core_api, volume_name):  # NO
...
+
+    1. Create a pod using Longhorn volume
+    2. Write some data to the volume then get the md5sum
+    3. Create a snapshot
+
+    Steps:
+
+    1. Delete the pod and wait for the volume detached
+    2. Randomly pick up a replica and manually delete or
+       empty the volume meta file in this replica data path
```
Comment (c3y1huang) [inline manager/integration/tests/test_basic.py:3666]: > Randomly pick up a replica and manually delete or empty the volume meta file in this replica data path
It seems we are testing for 2 cases here. Suggest separating them.

Comment (mantissahz) [reply manager/integration/tests/test_basic.py]: Yeah, separate it into two cases `"file does not exist"` and `"file is empty"`. Thanks.

**Comment by mantissahz** [comment]
> Can we add another case to delete volume meta when the volume is attached and having some IO?

Add test case `test_volume_metafile_deleted_when_writing_data`
Metadata file would be reconstructed after detaching the volume.

File: manager/integration/tests/test_basic.py [hunk 3]
```diff
@@ -3642,6 +3643,252 @@ def test_volume_toomanysnapshots_condition(client, core_api, volume_name):  # NO
...
+                               core_api,
+                               volume_name,
+                               pod,
+                               test_pod_name,
+                               data_path1,
+                               data_md5sum1,
+                               data_path2, snap1)
+
+
+def test_volume_metafile_deleted_when_writing_data(client, core_api, volume_name, csi_pv, pvc, pod, pod_make):  # NOQA
```
Comment (c3y1huang) [inline manager/integration/tests/test_basic.py:3812]: There seem to be duplicated codes to `prepare_data_volume_metafile`. Is it possible to add the data writing part to the  `prepare_data_volume_metafile` trigger the logic via the function argument?
```
    # create random data to the pod in the background
    command = 'dd if=/dev/urandom of=' + data_path1 + ' bs=1M' +\
              ' count=' + str(common.DATA_SIZE_IN_MB_4) + ' &'
    exec_command_in_pod(core_api, command, test_pod_name, 'default')
```

Same for the result validation part.

File: manager/integration/tests/test_basic.py [hunk 4]
```diff
@@ -3642,6 +3643,252 @@ def test_volume_toomanysnapshots_condition(client, core_api, volume_name):  # NO
...
+
+    # delete the volume metadata file of the volume on this host
+    command = ["rm", "-f", volume_meta_file]
+    subprocess.check_call(command)
+
+    # make the volume detached and metafile should be reconstructed.
+    delete_and_wait_pod(core_api, test_pod_name)
+    wait_for_volume_detached(client, volume_name)
+
+    # recreate the pod and check if the volume will become healty
```
Comment (c3y1huang) [inline manager/integration/tests/test_basic.py:3873]:
```suggestion
    # recreate the pod and check if the volume will become healthy
```

**Review by c3y1huang — CHANGES_REQUESTED** [review]

File: manager/integration/tests/test_basic.py [hunk 5]
```diff
@@ -3642,6 +3643,252 @@ def test_volume_toomanysnapshots_condition(client, core_api, volume_name):  # NO
...
+        },
+    }]
+
+    snap1 = create_snapshot(client, volume_name)
+
+    # choose the replica that is on this host
+    volume = client.by_id_volume(volume_name)
+    host_id = get_self_host_id()
+    for replica in volume.replicas:
+        if replica.hostId == host_id:
```
[the remark this answers is not in the document]
Comment (mantissahz) [reply manager/integration/tests/test_basic.py:3679]: Added checking `assert replica is not None`.
Thanks.

File: manager/integration/tests/test_basic.py [hunk 6]
```diff
@@ -3642,6 +3643,252 @@ def test_volume_toomanysnapshots_condition(client, core_api, volume_name):  # NO
...
+
+    return snap1, volume_meta_file, test_pod_name, data_md5sum1
+
+
+def rw_snapshot_volume_testing(client, core_api, volume_name, pod, test_pod_name, data_path1, data_md5sum1, data_path2, snap): # NOQA
+    """
+    Test volume I/O and take/delete a snpashot
+    """
+
+    # recreate the pod and check if the volume will become healty
```
[the remark this answers is not in the document]
Comment (mantissahz) [reply manager/integration/tests/test_basic.py]: Modified. Thanks.

File: manager/integration/tests/test_basic.py [hunk 7]
```diff
@@ -3642,6 +3643,252 @@ def test_volume_toomanysnapshots_condition(client, core_api, volume_name):  # NO
...
+    volume_meta_file = replica_data_path + "/volume.meta"
+
+    # make the volume detached
+    delete_and_wait_pod(core_api, test_pod_name)
+    wait_for_volume_detached(client, volume_name)
+
+    return snap1, volume_meta_file, test_pod_name, data_md5sum1
+
+
+def rw_snapshot_volume_testing(client, core_api, volume_name, pod, test_pod_name, data_path1, data_md5sum1, data_path2, snap): # NOQA
```
[the remark this answers is not in the document]
Comment (mantissahz) [reply manager/integration/tests/test_basic.py]: Modified. Thanks.

File: manager/integration/tests/test_basic.py [hunk 8]
```diff
@@ -3642,6 +3643,252 @@ def test_volume_toomanysnapshots_condition(client, core_api, volume_name):  # NO
...
+
+
+def test_volume_metafile_deleted(client, core_api, volume_name, csi_pv, pvc, pod, pod_make):  # NOQA
+    """
+    Test volume metafile deleted
+
+    Context:
+
+    If the volume still works fine when there is a invalid volume meta file
+    in the replica data path.
```
[the remark this answers is not in the document]
Comment (mantissahz) [reply manager/integration/tests/test_basic.py]: Modified. Thanks.

File: manager/integration/tests/test_basic.py [hunk 9]
```diff
@@ -3642,6 +3643,252 @@ def test_volume_toomanysnapshots_condition(client, core_api, volume_name):  # NO
...
+
+
+def test_volume_metafile_empty(client, core_api, volume_name, csi_pv, pvc, pod, pod_make):  # NOQA
+    """
+    Test volume metafile empty
+
+    Context:
+
+    If the volume still works fine when there is a invalid volume meta file
+    in the replica data path.
```
[the remark this answers is not in the document]
Comment (mantissahz) [reply manager/integration/tests/test_basic.py]: Modified. Thanks.

File: manager/integration/tests/test_basic.py [hunk 10]
```diff
@@ -3642,6 +3643,252 @@ def test_volume_toomanysnapshots_condition(client, core_api, volume_name):  # NO
...
+
+    Steps:
+
+    1. Empty volume meta file in this replica data path
```
[the remark this answers is not in the document]
Comment (mantissahz) [reply manager/integration/tests/test_basic.py]: Modified. Thanks.

File: manager/integration/tests/test_basic.py [hunk 11]
```diff
@@ -3642,6 +3643,252 @@ def test_volume_toomanysnapshots_condition(client, core_api, volume_name):  # NO
...
+    # 1 snap1 1 snap2 1 volume-head
+    wait_for_snapshot_count(volume, 3)
+    volume.snapshotDelete(name=snap.name)
+    # 1 snap2 1 volume-head
+    wait_for_snapshot_count(volume, 2)
+
+
+def test_volume_metafile_deleted(client, core_api, volume_name, csi_pv, pvc, pod, pod_make):  # NOQA
+    """
+    Test volume metafile deleted
```
[the remark this answers is not in the document]
Comment (mantissahz) [reply manager/integration/tests/test_basic.py]: Removed.Thanks.

File: manager/integration/tests/test_basic.py [hunk 12]
```diff
@@ -3642,6 +3643,252 @@ def test_volume_toomanysnapshots_condition(client, core_api, volume_name):  # NO
...
+
+
+def test_volume_metafile_deleted_when_writing_data(client, core_api, volume_name, csi_pv, pvc, pod, pod_make):  # NOQA
+    """
+    Test volume metafile deleted
+
+    Context:
+
+    If the volume still works fine when there is a invalid volume meta file
+    in the replica data path.
```
[the remark this answers is not in the document]
Comment (mantissahz) [reply manager/integration/tests/test_basic.py]: Modified. Thanks.

File: manager/integration/tests/test_basic.py (hunk 3 above)
Comment (mantissahz) [reply manager/integration/tests/test_basic.py:3822]: Merged into the `prepare_data_volume_metafile`. Thanks.

File: manager/integration/tests/test_basic.py (hunk 4 above)
Comment (mantissahz) [reply manager/integration/tests/test_basic.py]: Modified. Thanks.

File: manager/integration/tests/test_basic.py [hunk 13]
```diff
@@ -3643,6 +3644,237 @@ def test_volume_toomanysnapshots_condition(client, core_api, volume_name):  # NO
...
+    snap1 = create_snapshot(client, volume_name)
+
+    # choose the replica that is on this host
+    volume = client.by_id_volume(volume_name)
+    host_id = get_self_host_id()
+    for replica in volume.replicas:
+        if replica.hostId == host_id:
+            break
+
+    err_mgs = "Host-ID: {}, volume replica: {}".format(host_id, replica)
```
Comment (c3y1huang) [inline manager/integration/tests/test_basic.py:3682]:
```suggestion
    err_msg = "Host-ID: {}, volume replica: {}".format(host_id, replica)
```

You can also use f-string instead.
```
f'Host-ID: {host_id}, volume replica: {replica}'
```

File: manager/integration/tests/test_basic.py [hunk 14]
```diff
@@ -3643,6 +3644,237 @@ def test_volume_toomanysnapshots_condition(client, core_api, volume_name):  # NO
...
+        data_path1,
+        data_md5sum1,
+        data_path2,
+        snap1
+    )
+
+
+def test_volume_metafile_deleted_when_writing_data(client, core_api, volume_name, csi_pv, pvc, pod, pod_make):  # NOQA
+    """
+    Test volume metafile deleted
```
Comment (c3y1huang) [inline manager/integration/tests/test_basic.py:3825]: Redundant.

File: manager/integration/tests/test_basic.py [hunk 15]
```diff
@@ -3643,6 +3644,237 @@ def test_volume_toomanysnapshots_condition(client, core_api, volume_name):  # NO
...
+        # make the volume detached
+        delete_and_wait_pod(core_api, test_pod_name)
+        wait_for_volume_detached(client, volume_name)
+
+    return snap1, volume_meta_file, test_pod_name, data_md5sum1
+
+
+def check_volume_and_snapshot_after_corrupting_volume_metadata_file(client, core_api, volume_name, pod, test_pod_name, data_path1, data_md5sum1, data_path2, snap): # NOQA
+    """
+    Test volume I/O and take/delete a snpashot
```
Comment (c3y1huang) [inline manager/integration/tests/test_basic.py:3704]:
```suggestion
    Test volume I/O and take/delete a snapshot
```

**Review by c3y1huang — CHANGES_REQUESTED** [review]

File: manager/integration/tests/test_basic.py (hunk 13 above)
Comment (mantissahz) [reply manager/integration/tests/test_basic.py]: use ```f'Host-ID: {host_id}, volume replica: {replica}'``` instead. Thanks.

File: manager/integration/tests/test_basic.py (hunk 14 above)
Comment (mantissahz) [reply manager/integration/tests/test_basic.py]: Removed. Thanks.

File: manager/integration/tests/test_basic.py (hunk 15 above)
Comment (mantissahz) [reply manager/integration/tests/test_basic.py]: Modified. Thanks

**Review by c3y1huang — APPROVED** [review]

**Comment by yangchiu** [comment]
@Mergifyio backport v1.4.x
